Support of .git folder and .gitignore

aime-risson opened the pull request:
As described in #82 aeweb-cli wasn't handeling git folders.
Therefore I added a `--git-folder` argument.

It works by parsing the .gitignore file and then only uploading files that do not match the gitignore patterns.

to test it simply deploy a git folder as usual but add the `--git-folder` argument.

apoorv-2204 commented:
![Screenshot_20221128_224212](https://user-images.githubusercontent.com/90304197/204339647-610334fd-fb7e-412c-abf8-d30a5a814a68.png)

aime-risson commented:
> path.relative() expects a folder rather than a file as its first argument.
> ![Screenshot_20221128_224212](https://user-images.githubusercontent.com/90304197/204339647-610334fd-fb7e-412c-abf8-d30a5a814a68.png)

Hi @apoorv-2204,
I do not get this error with or without the `--git-folder` command. 

My folder also includes a .git/HEAD file and it works fine.
Maybe it is an os based error? 
Do you have more info on where the error really occurs?
Thanks

apoorv-2204 commented:
> > path.relative() expects a folder rather than a file as its first argument.
> > ![Screenshot_20221128_224212](https://user-images.githubusercontent.com/90304197/204339647-610334fd-fb7e-412c-abf8-d30a5a814a68.png)
> 
> Hi @apoorv-2204, I do not get this error with or without the `--git-folder` command.
> 
> My folder also includes a .git/HEAD file and it works fine. Maybe it is an os based error? Do you have more info on where the error really occurs? Thanks

Okay will retry.

apoorv-2204 commented:
> > path.relative() expects a folder rather than a file as its first argument.
> > ![Screenshot_20221128_224212](https://user-images.githubusercontent.com/90304197/204339647-610334fd-fb7e-412c-abf8-d30a5a814a68.png)
> 
> Hi @apoorv-2204, I do not get this error with or without the `--git-folder` command.
> 
> My folder also includes a .git/HEAD file and it works fine. Maybe it is an os based error? Do you have more info on where the error really occurs? Thanks

Can you please share your node js version ?

aime-risson commented:
> > > path.relative() expects a folder rather than a file as its first argument.
> > > ![Screenshot_20221128_224212](https://user-images.githubusercontent.com/90304197/204339647-610334fd-fb7e-412c-abf8-d30a5a814a68.png)
> > 
> > 
> > Hi @apoorv-2204, I do not get this error with or without the `--git-folder` command.
> > My folder also includes a .git/HEAD file and it works fine. Maybe it is an os based error? Do you have more info on where the error really occurs? Thanks
> 
> Can you please share your node js version ?

I'm using node js v18.11.0

apoorv-2204 commented:
> > > path.relative() expects a folder rather than a file as its first argument.
> > > ![Screenshot_20221128_224212](https://user-images.githubusercontent.com/90304197/204339647-610334fd-fb7e-412c-abf8-d30a5a814a68.png)
> > 
> > 
> > Hi @apoorv-2204, I do not get this error with or without the `--git-folder` command.
> > My folder also includes a .git/HEAD file and it works fine. Maybe it is an os based error? Do you have more info on where the error really occurs? Thanks
> 
> Can you please share your node js version ?
> 
> @aime-risson ![Screenshot_20221202_121041](https://user-images.githubusercontent.com/90304197/205231375-4f80765b-6e87-493c-aa70-e39f59625109.png)
> 
> ![Screenshot_20221202_121127](https://user-images.githubusercontent.com/90304197/205231432-3e51eb3a-ba03-4c69-92a2-4009d63d725c.png)

May be seems a issue with ignore dependency.

apoorv-2204 commented:
this code resolves the issue of relative path.
@aime-risson 
```
else {
    // check if file's pattern corresponds to gitignore patterns
    const absolutePath = getAbsolutePath(entry)
    if (!gitignore.ignores(getAbsolutePath(entry))) {
      handleFile(entry, files);
    }
  }
}

// ../path -> path
// /path -> path
function getAbsolutePath(entryPath) {
  var absolutePath = path.join(entryPath)

  if (absolutePath.startsWith(`..${path.sep}`)) return absolutePath.substr(3)
  if (absolutePath.startsWith(`${path.sep}`)) return absolutePath.substr(1)

  return absolutePath;
}
```

aime-risson commented:
Added the function and it works well !

I just changed the line `if (!gitignore.ignores(getAbsolutePath(entry)))` to `if (!gitignore.ignores(absolutePath))` as we had just calculated it already on the line above.
@apoorv-2204

apoorv-2204 commented:
> Added the function and it works well !
> 
> I just changed the line `if (!gitignore.ignores(getAbsolutePath(entry)))` to `if (!gitignore.ignores(absolutePath))` as we had just calculated it already on the line above. @apoorv-2204

my bad Great work!!

samuelmanzanera commented on commands/cli.js:
```diff
@@ -26,11 +28,24 @@ export function normalizeFolderPath(folderPath) {
+  
+  // if the folder is git folder then we should ignore 
+  // .git and files / folders in gitignore
+  if (isGitFolder){
```
We might do the reverse operation.
Instead of expecting this flag, we might look at the existence of the .gitignore file or .git folder to not deploy by default the folder attached.
But if we want to deploy the files referenced in the .gitignore, add a flag to explicit it (`--include-git-ignored-files`) 
In all cases, there is no point to deploy the .git folder

aime-risson replied:
> We might do the reverse operation.
> Instead of expecting this flag, we might look at the existence of the .gitignore file or .git folder to not deploy by default the folder attached.
> But if we want to deploy the files referenced in the .gitignore, add a flag to explicit it (`--include-git-ignored-files`)
> In all cases, there is no point to deploy the .git folder

Hadn't seen your comment sorry, should be fixed now. `.git` is always ignored, `.gitignore` files are ignored by default unless  `--include-git-ignored-files` flag is added to the deploy command.

aime-risson reviewed (commented):
Fix conflicts

apoorv-2204 commented:
> Fix conflicts

https://github.com/archethic-foundation/aeweb-cli/pull/93

Hi @aime-risson , I was unable to push changes on your branch, I created a new pr, with merge of your branch into the new branch. With that your external contribution is still recorded!!.Will also mention as author on last commit,

Cheers 😄

aime-risson commented:
> > Fix conflicts
> 
> 
> 
> https://github.com/archethic-foundation/aeweb-cli/pull/93
> 
> 
> 
> Hi @aime-risson , I was unable to push changes on your branch, I created a new pr, with merge of your branch into the new branch. With that your external contribution is still recorded!!.Will also mention as author on last commit,
> 
> 
> 
> Cheers 😄 

Great thanks!

Neylix commented:
Related PR #93 has been merged in main branch. Closing this one
